src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs

```diff
@@ -41,16 +42,28 @@ public DecompilationMetadataAsSourceFileProvider()
             Location? navigateLocation = null;
             var topLevelNamedType = MetadataAsSourceHelpers.GetTopLevelContainingNamedType(symbol);
             var symbolId = SymbolKey.Create(symbol, cancellationToken);
-            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
 
             // If we've been asked for signatures only, then we never want to use the decompiler
             var useDecompiler = !signaturesOnly && allowDecompilation;
+
+            // If the assembly wants to suppress decompilation we respect that
             if (useDecompiler)
             {
                 useDecompiler = !symbol.ContainingAssembly.GetAttributes().Any(attribute => attribute.AttributeClass?.Name == nameof(SuppressIldasmAttribute)
                     && attribute.AttributeClass.ToNameDisplayString() == typeof(SuppressIldasmAttribute).FullName);
             }
 
+            var refInfo = GetReferenceInfo(compilation, symbol.ContainingAssembly);
+
+            // If its a reference assembly we won't get real code anyway, so better to
+            // not use the decompiler, as the stubs will at least be in the right language
+            // (decompiler only produces C#)
+            if (useDecompiler)
+            {
+                useDecompiler = !refInfo.isReferenceAssembly;
```

Comment:
This is the crux of the fix: Deciding not to use the decompiler _before_ we decide which language we're going to display.